### ISSUE TITLE
dnscrypt-proxy IPv6 Windows 10 problems

### ISSUE THREAD

**Issue by Zenith-Nadir** [jedisct1/dnscrypt-proxy#635]
Hi, 
I want to use dnscrypt-proxy with IPv6 on Windows 10 x64. I have followed your Configuration guide, however I am encountering multiple problems with IPv6:
1. When I test C:\Program Files (x86)\bitbeans\Simple DNSCrypt\dnscrypt-proxy>.\dnscrypt-proxy.exe dnscrypt-proxy.conf with IPv4 settings, everything succeeds, but when I change to IPv6 I am not able to connect by any means. Please, see below:
2. When I change config to LocalAddress [::FFFF:127.0.0.1]:53, I get following error: [ERROR] Unable to bind (UDP) [Cannot assign requested address [WSAEADDRNOTAVAIL ]], despite that I am running COMMAND with elevated privilages.
3. When I change config to LocalAddress ::FFFF:127.0.0.1:53 (without brackets), I am getting: syntax error line 70: [:FFFF:127.0.0.1:53]
4. When I change config to LocalAddress [::1]:53, and invoke any IPv4 resolver, everything seems to work:
04/22/17 14:22:00 [INFO] + DNS Security Extensions are supported
04/22/17 14:22:00 [INFO] + Namecoin domains can be resolved
04/22/17 14:22:00 [INFO] + Provider supposedly doesn't keep logs
04/22/17 14:22:00 [NOTICE] Starting dnscrypt-proxy 1.9.4
04/22/17 14:22:00 [INFO] Loading plugin [C:\Program Files (x86)\bitbeans\Simple DNSCrypt\dnscrypt-proxy\libdcplugin_example_cache.dll]
04/22/17 14:22:00 [INFO] Loaded plugin: [A basic DNS cache]
04/22/17 14:22:00 [INFO] Generating a new session key pair
04/22/17 14:22:00 [INFO] Done
04/22/17 14:22:00 [INFO] Server certificate with serial #12345 received (removed for security reasons)
04/22/17 14:22:00 [INFO] This certificate is valid
04/22/17 14:22:00 [INFO] Server key fingerprint is XXXX: (removed for security reasons)
04/22/17 14:22:00 [NOTICE] Proxying from [::1]:53 to XXX.XXX.XXX.XXX:PORT (removed for security reasons)
but
are not resolving IPv6 addresses. They do resolve names to IPv4 addresses, but not IPv6 addresses or IPv6 names.

4. a) When I change config to LocalAddress [::1]:53, and invoke any IPv6 resolver, I get every time (with multiple different resolvers tried), following error:
04/22/17 14:25:48 [INFO] + Provider supposedly doesn't keep logs
04/22/17 14:25:48 [NOTICE] Starting dnscrypt-proxy 1.9.4
04/22/17 14:25:48 [INFO] Loading plugin [C:\Program Files (x86)\bitbeans\Simple DNSCrypt\dnscrypt-proxy\libdcplugin_example_cache.dll]
04/22/17 14:25:48 [INFO] Loaded plugin: [A basic DNS cache]
04/22/17 14:25:48 [INFO] Generating a new session key pair
04/22/17 14:25:48 [INFO] Done
04/22/17 14:26:03 [ERROR] Unable to retrieve server certificates
04/22/17 14:26:04 [INFO] Refetching server certificates
04/22/17 14:26:19 [ERROR] Unable to retrieve server certificates
04/22/17 14:26:22 [INFO] Refetching server certificates
And it never connects. Despite that its corresponding IPv4 server connects instantly.
I tried also to manually specify from .cvs file
ProviderName 
ProviderKey
ResolverAddress
also to no avail.

I am wondering what am I doing wrong, or what I set up wrong or did not set up, but should in config file.
I have well working IPv6 and have no issues with it otherwise, but I am unable to get it working with dnscrypt-proxy.
Any advice will be greatly appreciated.

Sincerely,
Zenith-Nadir

**Comment by jedisct1** [comment]
`[::FFFF:127.0.0.1]` is just a different way to write `127.0.0.1`, i.e. this is an IPv4 address. If you already have something listening to `127.0.0.1`, you can't have something else listening to the same IP.

You don't need to have the proxy listen to a local IPv6 address if it's already listening to an IPv4 address. Just use `[::FFFF:127.0.0.1]` as the resolvers address to use for IPv6 queries in the Windows settings.

The resolvers themselves don't care about the type of IP address clients connect to. They will respond to "what's the IPv6 address for that name" even if queries are sent using IPv4. So, if a resolver is reachable using both IPv4 and IPv6, using IPv6 is often just slower and less reliable.

`Unable to retrieve server certificates` means that a regular, standard, non-encrypted DNS query sent to that resolver never got a response. This happens even before anything is sent using DNSCrypt. Check that your router/firewall/antivirus properly support IPv6 and don't block port 443.

**Comment by Zenith-Nadir** [comment]
Hello jedisct1,
Thank you for such a quick response, and your interest to assist with my issues. THANKS!
Regarding ::FFFF:127.0.0.1, I know it is an old transition IPv4 to IPv6 notation.
My machine responds to pings to ::FFFF:127.0.0.1 and displays 127.0.0.1
Otherwise netstat -an does not show anything listening on ::FFFF:127.0.0.1 address, neither ::1, there are several instances listening on 127.0.0.1, but none on port 443, only one on port 53 while running dnscrypt-proxy, and none without dnscrypt-proxy running, but this does not seem to cause any problems anyway.
When I test dnscrypt-proxy using LocalAddress 127.0.0.1:53, I get the same results: IPv4 responses are instant from responsers using :443 port (nothing blocked), but when I specify any IPv6 responders all of them are Unable to retrieve server certificates.
Am I missing something? If I query IPv4 or IPv6 responders on 127.0.0.1 shouldn't they respond in IPv4?
Also, I specifically chosen :443 specified responters and IPv4 do respond, and all IPv6 are: Unable to retrieve server certificates.
Where am I making mistake?
I will try to disable all firewalls, but this will have to wait till tomorrow or later.
Thank you very much again for your assistance.
Sincerely,
Zenith-Nadir

**Comment by Zenith-Nadir** [comment]
Hello jedisct1,
It appears it was my router's ip6tables firewall allowing only udp port 53 for DNS.
Turns out that I had to permit not only port 443, but any port remote resolver is listening to.
Since, this does not seem to be any standard, normalized port, and there are resolvers listening on different ports (1053, 2053, 5353, 5443, etc), I had to allow those ports to be able to get resolvers to connect.
Thanks for your assistance and hints.
This would be very helpful for troubleshooting, if mentioned in Configuration github web page, and not only 443 port, but others, too. Especially for strict ip6tables or other firewalls.
Thanks again
Very Sincerely,
Zenith-Nadir
